src/genotyper.cc

```diff
@@ -41,6 +41,286 @@ static bool is_pseudo_ref_record(const bcf_hdr_t* hdr, bcf1_t* record) {
     return true;
 }
 
+class IFormatFieldHelper {
+
+public:
+    // basic information for the retained field
+    retained_format_field field_info;
+
+
+    // Expected number of samples in output bcf record
+    int n_samples;
+
+    // Expected number of values per sample in the **output** (ie unified site)
+    // Note this may differ from the count of input for RetainedFieldNumber::ALLELES
+    // and RetainedFieldNumber::ALT since the number of alleles may differ in input and
+    // output
+    int count;
+
+     IFormatFieldHelper(const retained_format_field field_info_, int n_samples_, int count_) : field_info(field_info_), n_samples(n_samples_), count(count_) {}
+
+     IFormatFieldHelper() = default;
+
+     virtual Status add_record_data(const string& dataset, const bcf_hdr_t* dataset_header, bcf1_t* record,
+                                    const map<int, int>& sample_mapping, const vector<int> allele_mapping,
+                                    const vector<string>& field_names={}, int n_val_per_sample=-1) = 0;
+
+     virtual Status update_record_format(const bcf_hdr_t* hdr, bcf1_t* record) = 0;
+
+     virtual ~IFormatFieldHelper() = default;
+
+};
+
+template <class T>
+class FormatFieldHelper : public IFormatFieldHelper {
+
+    // Boolean on whether the AD field has been handled
+    bool handled_AD_field = false;
+
+    // A vector of length n_samples, where each element
+    // is a vector consisting of format field value from
+    // record(s) related to the given sample
+    vector<vector<T>> format_v;
```

Comment:
The top level length is n_samples * count, and the second layer is up to n_records... added in more comment explanations